### PR TITLE
fix(ledger): bind spend authorization to the minted transfer id (#985)

### PR DESCRIPTION
**Pull request by ewindisch**
Closes the N1 pre-activation gate from #985 (fable review of PR #981).

## The gap

`admit()` verified the spend authorization's signature, grant binding, host, and amount ceiling — but never compared `authz.transfer_id` to the transfer id minted for this admission, even though that field is folded into the signed digest. One signed authorization therefore authenticated arbitrarily many distinct transfers (different nonces), each up to `max_amount`, for its whole `exp` window: replay amplification within the subject's own credit and a break of the per-transfer authorization semantics (#681).

## The fix

- Mint the deterministic transfer id **before** authz verification (it is a pure hash of `grant_cid` + `nonce`, no side effects) and hard-reject on `authz.transfer_id != transfer_id`.
- Add the negative test the happy path was masking: `authz_for_different_transfer_denied_hard` signs for nonce 7 and submits nonce 8 — signature verifies, minted id differs, admission must hard-reject with `InvalidSpendAuthorization`.

## Verification

All 19 `services::ledger` tests pass under `--features ledger`, including the pre-existing roundtrip (proving the equality check does not break the legitimate single-nonce flow).

## Scope

Deliberately N1-only. N2–N6 from the same review (grant cache not generation-scoped, PQ downgrade labeling, post_actual remainder release, factory block_on, optional hyprstream-crypto) remain open on #985 — N2 was independently re-confirmed as a second activation blocker by the 2026-07-21 UCAN LOE evaluation and should be its own change.

Refs #985, #925, #681. Consumer context: www-cyberdione-ai epic #39 P3 is hard-gated on this landing before any entitlement enforcement activates.

### CODE REVIEW

**Comment by ewindisch** [comment]
Independent review (gpt-5.6-sol, exact-head f4b6f46be) returned REVISE with two blocking findings, both now addressed in fec6f970d: (1) zero-amount admissions bypassed the entire authz block — authorization is now required for every admission; (2) `authz.exp` was signed but never enforced — now fail-closed at the boundary. Negative tests added for both; 21/21 ledger tests green. Finding 3 (subject↔grant-holder binding) predates this patch and is filed as its own activation blocker. Re-review dispatched at the new head.

**Comment by ewindisch** [comment]
## CodeQL triage: 10 alerts, all false positives — fixed by restructuring

**Alerts:** `rust/hard-coded-cryptographic-value` ("hard-coded value used as a nonce"), 10 instances, all in `crates/hyprstream/src/services/ledger/enforcer.rs` lines 541/565/596/599/602/631/657/663/686/720.

**Flagged expressions:** every instance is a literal `u128` (`nonce: 1`, `nonce: 7`, `nonce: 8`, `nonce: 9`, and the `transfer_nonce` argument literals `1`/`5`/`7` passed to the local test helper `signed_authz(...)`) inside `#[cfg(test)] mod tests` (module starts at line 416).

### FP vs real — verified, not assumed

Read the production path the rule points at:

- `AdmissionRequest::nonce` and `signed_authz`'s `transfer_nonce` parameter both feed `mint_transfer_id(grant_cid, nonce) -> TransferId`, which is `blake3("hs-ledger-xfer-id-v1" || grant_cid || nonce_be)` — a **public, deterministic idempotency-key derivation**, not a secret. There is no AEAD/IV usage anywhere in this module (grepped for `Nonce|IV|AeadCore|aes|chacha` in `services/ledger/*` — zero real hits). The ledger's crypto here is Ed25519 + ML-DSA-65 **signing** over a domain-separated BLAKE3 digest ([`SpendAuthorization::digest`](crates/hyprstream/src/services/ledger/credit_gate.rs)), which takes no nonce input at all.
- CodeQL's nonce-role heuristic pattern-matches on the parameter name/role (`nonce`), not on whether the value is actually security-critical secret material. Same root cause as #1121 (test literals flagged as hard-coded crypto values despite being public, non-secret idempotency inputs never reaching a real crypto sink).
- Checked for the #1111 pattern (FP masking a real adjacent bug): no adjacent nonce-handling bug found — the actual `#985` binding logic (`authz.transfer_id != transfer_id` in `LocalEnforcer::admit`) is sound and exercised directly by `authz_for_different_transfer_denied_hard`, which passed before and after this change.

**Determination: false positive**, structurally identical to #1121.

### Fix: restructured, not suppressed

Per repo convention (#1111/#1121): added a `rand_nonce()` test helper drawing 128 bits from `OsRng` and replaced all 10 flagged literals with it, preserving each test's relational nonce semantics:
- `authz_for_different_transfer_denied_hard` now draws two independent nonces and asserts them distinct (was: literal `7` vs `8`).
- `expired_authz_denied_hard`, `admitted_then_post_and_void_roundtrip`, `insufficient_credit_rejects_retryable_not_queues` now share one `let nonce = rand_nonce();` between the signing call and the request (was: same literal repeated).
- `anonymous_subject_denied_hard`, `zero_amount_without_authz_denied_hard` just draw a fresh nonce (was: arbitrary literal, value never asserted on).

This clears the alert at the root rather than suppressing it, and is at least as faithful a test of production behavior (client nonces are attacker/caller-supplied entropy, not fixed constants).

### Item 4 — verification-side binding sanity check (#985's actual subject)

Confirmed the fix binds spend authorization to the transfer id **on the verification side**, not only at mint:

```rust
// LocalEnforcer::admit, verification path:
let transfer_id = mint_transfer_id(&req.grant_cid, req.nonce); // freshly minted from THIS request
...
if authz.transfer_id != transfer_id {
    return AdmissionResult::Rejected(Rejection::hard(
        DenyReason::InvalidSpendAuthorization("authz transfer id does not match this transfer".to_owned()),
    ));
}
```

`transfer_id` is minted independently from the incoming request's own `(grant_cid, nonce)` and compared against the signed `authz.transfer_id` (which is itself in the signed BLAKE3 digest — `SpendAuthorization::digest`). This is a real equality check on the verifier's own recomputation, not just an assertion the mint-time code happened to embed the right value. No gap found.

Also checked whether there's a live call site where `transfer_id` could be substituted between `admit()` returning `Admitted` and `reserve_durable` consuming it: there isn't one yet — per the module's own doc comment, `LocalEnforcer::admit` isn't wired into the `hyprstream-workers` scheduler PEP path yet (blocked on the open #761 group-authority decision, lands behind the `ledger` flag, default off). So the only production caller of `reserve_durable` is fed the same `transfer_id` `admit()` already validated.

### Verification

- `cargo build -p hyprstream --features ledger`: clean
- `cargo test -p hyprstream --lib --features ledger services::ledger`: 21/21 pass
- `cargo clippy -p hyprstream --all-targets --features ledger -- -D warnings`: clean

(`cargo fmt --check` fails on a clean `main` per #1140 — unrelated to this PR.)

Not merging — leaving that to a maintainer.

File: crates/hyprstream/src/services/ledger/enforcer.rs [hunk 1]
```diff
@@ -200,10 +200,22 @@ impl LocalEnforcer {
...
+
+        // 6. Spend-authorization verification (§5.3). Required for EVERY
+        //    admission, including zero-amount metered-rate jobs: `Admitted` is
+        //    the decision that permits work to start, and a zero reservation
+        //    cannot be converted into a nonzero post later — so an unsigned
+        //    zero admission would be unauthenticated work, not harmless
+        //    accounting (PR #1129 review, finding 1). A Classical-only client
+        //    passes no PQ key and is verified under the Classical policy
+        //    (labeled, not rejected).
+        {
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/services/ledger/enforcer.rs]: Verified REAL — escalating, not resolving (PQ / Hybrid-signing posture).

`enforcer.rs:269`:
```rust
let require_pq = self.require_pq && req.subject_pq_vk.is_some();
```
flows into `verify_spend_authz` (`credit_gate.rs:131-141`) → `verify_composite(..., require_pq)`. With `require_pq=false`, `verify_composite` performs **classical-only** verification (the PQ signature is not required). So when the enforcer is configured Hybrid (`self.require_pq == true`) but the request presents `subject_pq_vk: None`, the `&& req.subject_pq_vk.is_some()` clause forces `require_pq=false`, the PQ arm is skipped, and the admission is verified with classical assurance only. A client can therefore shed the PQ protection of a Hybrid ledger policy simply by omitting its PQ key — a silent downgrade that this repo has explicitly ratified *against*: under `Hybrid`/mandatory-hybrid, a missing PQ key must **fail closed**, never downgrade to classical (the mandatory-hybrid posture landed in #556/#1087; see also `CLAUDE.md` "CryptoPolicy under Hybrid: a missing PQ key fails closed at sign time — it never silently downgrades to classical").

The author comment at L205-208 frames `None` as "a Classical-only client … verified under the Classical policy (labeled, not rejected)." Whether a classical fallback is acceptable for *spend authorization under a Hybrid ledger policy* is precisely the crypto-policy call that needs an owner decision (fail-closed vs. labeled-admit), not a triage resolution. Keeping `require_pq` tied directly to `self.require_pq` and rejecting `subject_pq_vk: None` when Hybrid is on is the fail-closed option the thread proposes.

Also confirmed: no regression test covers the `self.require_pq=true` ∧ `req.subject_pq_vk=None` combination. Existing `verify_spend_authz` tests (`credit_gate.rs:504,508`) pass `Some(&pq_vk)` with `require_pq=true`; the enforcer-level tests that set `subject_pq_vk: None` (L592, L663) run under `require_pq_signatures: false` (L535). The downgrade path is untested.

**Comment by ewindisch** [comment]
## CodeQL reviewer follow-up — passing check still inherits 3 open alerts

The current head `abfd2a5bd` correctly removed the newly introduced test-literal alerts #101–#107 and five inherited alerts, and the production fix now recomputes `transfer_id = mint_transfer_id(grant_cid, req.nonce)` before comparing it with the signed `authz.transfer_id`.

However, the default branch still has these open alerts in the same file, and the PR head retains the exact literals:

- #83, `enforcer.rs:640`: `5` in `mint_transfer_id(&c, 5)`
- #84, `enforcer.rs:641`: `5` in `mint_transfer_id(&c, 5)`
- #85, `enforcer.rs:643`: `6` in `mint_transfer_id(&c, 6)`

They are test-only false positives, but at the analyzed/current `main` production revision they are **MASKING-SOMETHING-ELSE**: `LocalEnforcer::admit` did not bind the signed authorization transfer ID to the request-derived ID, allowing one valid authorization to amplify across different request nonces. This PR fixes that real adjacent defect.

Please finish the root restructuring in `transfer_id_is_deterministic_and_distinct_for_post_void`: draw two distinct `rand_nonce()` values, use the first for both `a` and `b`, and the second for `other`. That preserves the relational assertion and clears #83–#85. No suppression is warranted. I made no branch/source edits.

**Comment by ewindisch** [comment]
## Thread triage — REAL finding for the owning lane

**`services/ledger/enforcer.rs:218` — the Hybrid crypto policy silently downgrades to classical when `subject_pq_vk: None`.**

A configured Hybrid policy that proceeds with classical-only verification because the post-quantum key is absent is the **"absent expected value disables the check"** shape: the requirement is expressed, the material is missing, and the check quietly weakens instead of refusing. It violates the fail-closed mandatory-hybrid posture that https://github.com/hyprstream/hyprstream/issues/556 established by removing the classical floor deliberately.

A configured PQ requirement with absent PQ material must **refuse**, never proceed. This is the same defect class that produced a CRITICAL on https://github.com/hyprstream/hyprstream/pull/1111 (an absent confidential epoch turning verification off and accepting arbitrary plaintext) and a HIGH on https://github.com/hyprstream/hyprstream/pull/1193 — and in both of those it was introduced by a fix for something else, so check the surrounding change for a second instance.

Not resolved — this touches key material and is escalated per the standing rule.

**Comment by ewindisch** [comment]
## Thread triage — REAL findings for the owning lane

Triage lane-A verified the open review thread against the head (`abfd2a5bd`, `fix/985-bind-transfer-id`). CodeRabbit is correct on the merits; this needs an owner decision, so it is escalated, not resolved.

- **`crates/hyprstream/src/services/ledger/enforcer.rs:269`** — `let require_pq = self.require_pq && req.subject_pq_vk.is_some();` flows into `verify_spend_authz` → `verify_composite(.., require_pq)`. When the enforcer is configured Hybrid (`self.require_pq == true`) but the request omits `subject_pq_vk`, the `&& req.subject_pq_vk.is_some()` clause forces `require_pq=false`, the PQ arm is skipped, and the spend authorization is verified with classical assurance only. A client can shed the PQ protection of a Hybrid ledger policy by omitting its PQ key — a silent downgrade. The L205-208 comment frames this as an intentional classical fallback ("labeled, not rejected"), but that conflicts with the repo's mandatory-hybrid / fail-closed-on-missing-PQ posture (#556/#1087).
  **Owner decision:** either keep `require_pq` tied directly to `self.require_pq` and reject `subject_pq_vk: None` when Hybrid is on (fail-closed), or explicitly ratify a labeled-classical fallback for this ledger path. No test currently covers the `require_pq=true` ∧ `subject_pq_vk=None` combination — add one for whichever behavior is chosen.

**Comment by ewindisch** [comment]
## Correction and operator ruling — this is stronger than a "downgrade"

My earlier triage comment called this a silent downgrade. That understates it. The actual line is:

```rust
let require_pq = self.require_pq && req.subject_pq_vk.is_some();
```

`self.require_pq` is the **deployment's policy**. `req.subject_pq_vk` is **caller-supplied**. Conjoining them means **the requester decides whether the operator's requirement applies**: present a PQ key and you are held to PQ; omit it and you are not. That is not a policy selecting classical for a classical peer — it is a configured requirement with a caller-controlled off switch.

The inline comment defends it as *"A Classical-only client passes no PQ key and is verified under the Classical policy (labeled, not rejected)"* — the assurance-axis approach, which is legitimate in principle. **But `grep Assurance::Classical` across `services/ledger/` returns nothing.** Nothing labels the admission and nothing downstream refuses a classical-labeled admission for a PQ-required operation. As written, "labeled, not rejected" is simply *not rejected*.

## Operator ruling (Erica, 2026-07-23)

> **PQ-hybrid is required on ALL deployments, and classical must fail closed on non-interop surfaces.**

With the surface taxonomy she gave:

| Surface | Posture |
|---|---|
| **hyprstream ↔ hyprstream**, `did:at9p`, internal RPC, **ledger admission** | **Always PQ-hybrid, no fallback** |
| Third-party interop — JWKS verification of foreign tokens, atproto federation with `did:plc` peers, browser clients | Classical permitted; per RFC, try every supplied `kid` of a **known** alg and **ignore unknown algs rather than failing** |

**Ledger spend authorization is category one.** It binds a signed authorization to a minted transfer id and checks `authz.host == self.cell_identity` — an internal cluster operation between hyprstream components admitting work on a cell. There is no third party on that path to interop with, so there is no classical case to accommodate.

## Required change

**Remove the caller-supplied conjunct entirely.** A deployment requiring PQ must reject an admission presenting no PQ material — `subject_pq_vk: None` is a rejection, not a policy selection. Since PQ-hybrid is required on all deployments, `self.require_pq` is not a meaningful toggle on this surface either; if it stays, it must not be satisfiable by omission.

Add a test that fails without the fix: an admission with `subject_pq_vk: None` under a PQ-requiring deployment must be **REJECTED**, and reverting the change must make that test pass again.

Note the enforcer has **no production callers yet** (gated on #761), so this is pre-live — which is both why it went unnoticed and why it is cheap to fix now.

Refs https://github.com/hyprstream/hyprstream/issues/556 (fail-closed mandatory-hybrid posture — classical floor removed deliberately), https://github.com/hyprstream/hyprstream/issues/1184 (the JWKS known-alg/unknown-alg rule, which is the *interop* half and is correct there).

**Comment by ewindisch** [comment]
Addressed in 200a2ac78.

- Removed `LocalEnforcer`'s `require_pq` toggle from ledger admission and removed the `require_pq` argument from `CreditGate::verify_spend_authz`; this internal hyprstream↔hyprstream surface now always requires PQ-hybrid verification. `LedgerConfig::require_pq_signatures` remains for checkpoint signing, but no longer controls this admission boundary.
- Corrected the stale “Classical-only / labeled, not rejected” documentation. A missing `subject_pq_vk` now reaches mandatory composite verification and hard-rejects.
- Added `missing_pq_key_denied_hard`, with the fixture using the default PQ-requiring configuration.

Validation:
- `buildq -- cargo test -p hyprstream --lib --features ledger services::ledger`
- `buildq -- cargo clippy -p hyprstream --all-targets --features ledger -- -D warnings`
- Foreground pre-commit: workspace release Clippy, passed.
- Mutation check: temporarily changed mandatory composite verification back to classical fallback; the new missing-key test failed, then restored the secure implementation and reran the ledger suite.

Not merged.

**Comment by ewindisch** [comment]
VERDICT: REVISE

Reviewed PR #1129 at exact pushed head `200a2ac7819af0a269b89d93406ee46f8411ffb4`
against current `main` `69eda14916c0870c16f101eb4147f637e09f2043`.

## Findings

### HIGH — The pushed head conflicts with `main` across both security-critical files

GitHub reports `mergeable: CONFLICTING` / `mergeStateStatus: DIRTY`, and
`git merge-tree` confirms content conflicts in both
`crates/hyprstream/src/services/ledger/credit_gate.rs` and
`crates/hyprstream/src/services/ledger/enforcer.rs`.

This is not a mechanical conflict that can safely receive a merge-queuing PASS.
Current `main` has replaced the request-supplied `subject`, `subject_ed_vk`, and
`subject_pq_vk` fields with an opaque `AuthenticatedSubject` derived from the
verified RPC envelope, and it binds that authenticated DID to
`VerifiedGrant::holder`. Choosing the PR side while resolving the conflict
would restore caller-supplied identity/key fields and discard the holder-binding
boundary. Choosing `main` unchanged would retain the now-invalid
all-deployments `require_pq_signatures` toggle.

Rebase onto current `main`, preserve `AuthenticatedSubject` and
`HolderMismatch` enforcement, and port the final PQ fix onto that shape:
ledger admission must always call composite verification with PQ required,
must reject `AuthenticatedSubject.pq_vk: None`, and must not retain a
configuration or caller-controlled classical escape. The rebased head needs a
fresh review before PASS.

## Security review of the exact pushed head

No source defect was found in the pushed head itself:

- The #985 binding is enforced in `LocalEnforcer::admit` on the verification
  path: `transfer_id` is deterministically re-minted from the request's
  `(grant_cid, nonce)`, compared to signed `authz.transfer_id`, and rejected
  before `CreditGate::try_hold`.
- `SpendAuthorization::digest` covers `grant_cid`, `host`, `transfer_id`,
  `max_amount`, and `exp`; the later composite-signature verification prevents
  tampering with values checked before signature verification.
- Every admission, including `amount == 0`, traverses authorization verification.
  Missing authorization and expired authorization hard-reject.
- The prior caller-controlled conjunct is gone. `LocalEnforcer` no longer stores
  `require_pq`, and `CreditGate::verify_spend_authz` passes literal `true` to
  `verify_composite`. That verifier rejects both an absent PQ anchor and a
  classical-only signature. No `Option::None` skip, `NotApplicable`, default,
  feature flag, or assurance-label substitute exists in the changed boundary.
- The prior verdict's separate holder-binding concern has since landed on
  `main`; it is exactly why the rebase resolution must preserve `main`'s
  authenticated-subject implementation.

## What the change intentionally breaks

The final PQ commit intentionally breaks every classical-only ledger admission,
including deployments whose existing `LedgerConfig.require_pq_signatures` is
false. A missing authenticated ML-DSA key and a classical-only composite now
hard-reject. This is the required operator policy for the non-interop ledger
surface. The legitimate PQ-hybrid path remains healthy: the full
admit → durable reserve → partial post test passes.

## Test causality and validation

All Cargo commands used `buildq`. The first test attempt stopped before affected
code compiled because libtorch was not selected; rerunning with the repository's
cached CUDA 13.0 libtorch and system OpenSSL environment succeeded.

- `cargo test -p hyprstream --lib --features ledger services::ledger -- --nocapture`:
  **22 passed, 0 failed, 0 ignored**. This includes wrong-transfer rejection,
  zero-amount authorization, expiry, missing-PQ rejection, insufficient-credit
  behavior, and the legitimate admit/reserve/post path.
- Mutation check in a disposable detached worktree: removed transfer-ID
  equality, restored the zero-amount bypass, removed authz expiry, and changed
  mandatory PQ to caller-key-presence selection while retaining all tests.
  `cargo test ... services::ledger::enforcer::tests` then produced exactly four
  failures: `authz_for_different_transfer_denied_hard`,
  `zero_amount_without_authz_denied_hard`, `expired_authz_denied_hard`, and
  `missing_pq_key_denied_hard`; each received `Admitted`. The other five tests,
  including the happy path, passed. The mutations were restored, the disposable
  worktree was verified clean, and it was removed.
- `cargo build -p hyprstream --features ledger`: **passed**.
- `cargo clippy -p hyprstream --all-targets --features ledger -- -D warnings`:
  **passed**.
- `git diff --check`: **passed**. The reviewed worktree remains clean at the
  exact PR head.
